### PR TITLE
Fix throwing NPE when table-type is defined without type-param

### LINKED ISSUE #34732: Using the table type descriptor snippet completion item throws NPE
**Description:**
When I select the following completion item `type TypeName table<>`, it throws the NPE below.

Screenshot 2022-01-19 at 13 05 44

<details>
  <summary>Stack trace. </summary>

```
Error - 1:06:23 PM] Operation 'textDocument/semanticTokens/full' failed! {uri: '/Users/tharushijayasekara/Documents/Test/test-beta6/main.bal', error: 'java.lang.NullPointerException, org.wso2.ballerinalang.compiler.tree.types.BLangTableTypeNode.accept(BLangTableTypeNode.java:82)'}
java.lang.NullPointerException
	at io.ballerina.compiler.api.impl.SymbolFinder.visit(SymbolFinder.java:1180)
	at org.wso2.ballerinalang.compiler.tree.types.BLangTableTypeNode.accept(BLangTableTypeNode.java:82)
	at io.ballerina.compiler.api.impl.SymbolFinder.lookupNode(SymbolFinder.java:253)
	at io.ballerina.compiler.api.impl.SymbolFinder.visit(SymbolFinder.java:335)
	at org.wso2.ballerinalang.compiler.tree.BLangTypeDefinition.accept(BLangTypeDefinition.java:116)
	at io.ballerina.compiler.api.impl.SymbolFinder.lookup(SymbolFinder.java:230)
	at io.ballerina.compiler.api.impl.BallerinaSemanticModel.lookupSymbol(BallerinaSemanticModel.java:379)
	at io.ballerina.compiler.api.impl.BallerinaSemanticModel.symbol(BallerinaSemanticModel.java:165)
	at org.ballerinalang.langserver.semantictokens.SemanticTokensVisitor.processSymbols(SemanticTokensVisitor.java:441)
	at org.ballerinalang.langserver.semantictokens.SemanticTokensVisitor.visit(SemanticTokensVisitor.java:181)
	at io.ballerina.compiler.syntax.tree.SimpleNameReferenceNode.accept(SimpleNameReferenceNode.java:41)
	at io.ballerina.compiler.syntax.tree.NodeVisitor.visitSyntaxNode(NodeVisitor.java:846)
	at io.ballerina.compiler.syntax.tree.NodeVisitor.visit(NodeVisitor.java:455)
	at io.ballerina.compiler.syntax.tree.TypeParameterNode.accept(TypeParameterNode.java:49)
	at io.ballerina.compiler.syntax.tree.NodeVisitor.visitSyntaxNode(NodeVisitor.java:846)
	at io.ballerina.compiler.syntax.tree.NodeVisitor.visit(NodeVisitor.java:451)
	at io.ballerina.compiler.syntax.tree.TableTypeDescriptorNode.accept(TableTypeDescriptorNode.java:50)
	at io.ballerina.compiler.syntax.tree.NodeVisitor.visitSyntaxNode(NodeVisitor.java:846)
	at org.ballerinalang.langserver.semantictokens.SemanticTokensVisitor.visit(SemanticTokensVisitor.java:307)
	at io.ballerina.compiler.syntax.tree.TypeDefinitionNode.accept(TypeDefinitionNode.java:62)
	at io.ballerina.compiler.syntax.tree.NodeVisitor.visitSyntaxNode(NodeVisitor.java:846)
	at org.ballerinalang.langserver.semantictokens.SemanticTokensVisitor.getSemanticTokens(SemanticTokensVisitor.java:106)
	at org.ballerinalang.langserver.semantictokens.SemanticTokensUtils.getSemanticTokens(SemanticTokensUtils.java:64)
	at org.ballerinalang.langserver.BallerinaTextDocumentService.lambda$semanticTokensFull$15(BallerinaTextDocumentService.java:624)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```
</details>

Once `type TypeName table<>;` is inserted, hovering over `table<>` throws the NPE below.
Screenshot 2022-01-19 at 13 20 50

<details>
  <summary>Stack trace. </summary>

```
[Error - 1:09:33 PM] Operation 'text/hover' failed! {uri: '/Users/tharushijayasekara/Documents/Test/test-beta6/main.bal', [139:21], error: 'java.lang.NullPointerException, org.wso2.ballerinalang.compiler.tree.types.BLangTableTypeNode.accept(BLangTableTypeNode.java:82)'}
java.lang.NullPointerException
	at io.ballerina.compiler.api.impl.SymbolFinder.visit(SymbolFinder.java:1180)
	at org.wso2.ballerinalang.compiler.tree.types.BLangTableTypeNode.accept(BLangTableTypeNode.java:82)
	at io.ballerina.compiler.api.impl.SymbolFinder.lookupNode(SymbolFinder.java:253)
	at io.ballerina.compiler.api.impl.SymbolFinder.visit(SymbolFinder.java:335)
	at org.wso2.ballerinalang.compiler.tree.BLangTypeDefinition.accept(BLangTypeDefinition.java:116)
	at io.ballerina.compiler.api.impl.SymbolFinder.lookup(SymbolFinder.java:230)
	at io.ballerina.compiler.api.impl.BallerinaSemanticModel.lookupSymbol(BallerinaSemanticModel.java:379)
	at io.ballerina.compiler.api.impl.BallerinaSemanticModel.symbol(BallerinaSemanticModel.java:153)
	at org.ballerinalang.langserver.hover.HoverUtil.getHover(HoverUtil.java:86)
	at org.ballerinalang.langserver.BallerinaTextDocumentService.lambda$hover$1(BallerinaTextDocumentService.java:170)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```
</details>

For as long as the completion item remains to be in the file, NPEs continue to be thrown for every text edit.
<details>
  <summary>Stack trace. </summary>

```
[Error - 12:40:02 PM] Operation 'textDocument/semanticTokens/full' failed! {uri: '/Users/tharushijayasekara/Documents/Test/test-beta6/main.bal', error: 'java.lang.NullPointerException'}
java.lang.NullPointerException

[Error - 12:40:09 PM] Operation 'textDocument/semanticTokens/full' failed! {uri: '/Users/tharushijayasekara/Documents/Test/test-beta6/main.bal', error: 'java.lang.NullPointerException'}
java.lang.NullPointerException

[Error - 12:40:10 PM] Operation 'textDocument/semanticTokens/full' failed! {uri: '/Users/tharushijayasekara/Documents/Test/test-beta6/main.bal', error: 'java.lang.NullPointerException'}
java.lang.NullPointerException

[Error - 12:40:10 PM] Operation 'textDocument/semanticTokens/full' failed! {uri: '/Users/tharushijayasekara/Documents/Test/test-beta6/main.bal', error: 'java.lang.NullPointerException'}
java.lang.NullPointerException

[Error - 12:40:14 PM] Operation 'textDocument/semanticTokens/full' failed! {uri: '/Users/tharushijayasekara/Documents/Test/test-beta6/main.bal', error: 'java.lang.NullPointerException'}
java.lang.NullPointerException

[Error - 12:41:51 PM] Operation 'textDocument/semanticTokens/full' failed! {uri: '/Users/tharushijayasekara/Documents/Test/test-beta6/main.bal', error: 'java.lang.NullPointerException'}
java.lang.NullPointerException

[Error - 12:41:51 PM] Operation 'textDocument/semanticTokens/full' failed! {uri: '/Users/tharushijayasekara/Documents/Test/test-beta6/main.bal', error: 'java.lang.NullPointerException'}
java.lang.NullPointerException

[Error - 12:41:51 PM] Operation 'textDocument/semanticTokens/full' failed! {uri: '/Users/tharushijayasekara/Documents/Test/test-beta6/main.bal', error: 'java.lang.NullPointerException'}
java.lang.NullPointerException
```
</details>

**Steps to reproduce:**
Mentioned in the description.

**Affected Versions:**
stage-swan-lake

### PR DESCRIPTION
**Pull request by dulajdilshan**
## Purpose
$title

Fixes #34732 

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by mohanvive — DISMISSED** [review]
It would be better to get a review feedback from compiler team as well. @gimantha @hasithaa

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 1]
```diff
@@ -573,6 +577,12 @@ public void visit(BLangTypeDefinition typeDefinition) {
             analyzeDef(typeDefinition.typeNode, env);
         }
 
+        if (typeDefinition.symbol == null) {
+            typeDefinition.symbol = Symbols.createTypeDefinitionSymbol(Flags.asMask(typeDefinition.flagSet),
+                    names.fromIdNode(typeDefinition.name), env.enclPkg.packageID, symTable.noType, env.scope.owner,
+                    typeDefinition.name.pos, getOrigin(typeDefinition.name.value));
+        }
+
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:585]: This looks hacky. Why do we need this? IIRC, to fix the issue, adding a null check in the symbol finder is enough.

**Comment by pubudu91** [comment]
And please tag the PR and the issue with the relevant milestone

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java (hunk 1 above)
Comment (dulajdilshan) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:585]: Without this, the table symbol will still be null and cannot perform symbol(), references() on it

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:585]: There's no point creating a type definition symbol if the type itself cannot be resolved, right? How is this handled for maps?

e.g.,
```ballerina
type TypeName map<>;
```

Didn't observe the NPE for this.

Comment (dulajdilshan) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:585]: Here we are not getting the NPE, but renaming and find references features are not working

Comment (dulajdilshan) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:585]: @mohanvive @nadeeshaan

Comment (nadeeshaan) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:585]: I believe @dulajdilshan 's idea is correct here. It is a common use-case that the user tends to find the references in these scenarios. Also, the go-to definition as well. Hence It is beneficial to handle this. +1

Comment (dulajdilshan) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:585]: As discussed (with @mohanvive @nadeeshaan and @pubudu91). This is a common cause https://github.com/ballerina-platform/ballerina-lang/issues/34706#issuecomment-1047592733 we have to handle generally. Therefore, we have decided to remove creating type-def symbol for this level.

**Review by mohanvive — APPROVED** [review]

**Pull request merged**